### PR TITLE
Feature: Allow keepalive to be disabled when set to 0

### PR DESCRIPTION
**Pull request by pryorda**
### Summary

Allow Engineers to disable keeaplive by setting upstream_keepalive to 0. In our infrastructure we did not want to have nginx workers keeping the backend connections alive that were to externally lb services. This fixed various connection resets for us. 

### Full changelog

* Implemented ability to disable upstream_keepalive

### Issues resolved

Possibly: #3287

### CODE REVIEW

**Comment by pryorda** [comment]
Let me know if I missed something. This is my first lua pull.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+# Would be awesome to see this use docker-compose and build a fully
+# functional test environment
+
+COPY . /build
+
+WORKDIR /build
+RUN apk add --no-cache openssl openssl-dev git
+RUN make dev && make test
+
```
Comment (thibaultcha) [inline Dockerfile]: Please remove unrelated changes from this PR, such as this Dockerfile and the Makefile rule added below, along with the README change. Thank you!

File: kong/templates/nginx_kong.lua [hunk 2]
```diff
@@ -66,7 +66,9 @@ upstream kong_upstream {
     balancer_by_lua_block {
         Kong.balancer()
     }
+> if upstream_keepalive > 0 then 
     keepalive ${{UPSTREAM_KEEPALIVE}};
+> end
```
Comment (thibaultcha) [inline kong/templates/nginx_kong.lua]: We would need you to provide a test for this behaviour. Nginx config template test can be found here: https://github.com/Kong/kong/blob/master/spec/01-unit/003-prefix_handler_spec.lua#L55

File: kong.conf.default [hunk 3]
```diff
@@ -213,6 +213,8 @@
                                  # preserved in the cache of each worker
                                  # process. When this number is exceeded, the
                                  # least recently used connections are closed.
+                                 # 0 will remove the entry conf nginx-kong.conf
+                                 # when using the default template
```
Comment (thibaultcha) [inline kong.conf.default]: Will you please update this description to something like:
```
A value of `0` will disable this behavior altogether,
forcing each upstream request to open a new connection.
```

`nginx-kong.conf` and "the default template" are implementation details that should not be mentioned here. Additionally, the proposed text is more explicit as to what the behavior is when the given value is `0`.

**Comment by pryorda** [comment]
@thibaultcha Should be good to go now. Let me know if I missed anything or did it wrong.

**Comment by thibaultcha** [comment]
@pryorda Manually merged to next with a few edits, thanks!

Will you be able to provide documentation for this change as well? Thank you!

**Comment by pryorda** [comment]
@thibaultcha Created here: https://github.com/Kong/docs.konghq.com/pull/827

Thanks
